detect_test.go

```diff
@@ -0,0 +1,233 @@
+		Expect(os.Unsetenv("COMPOSER")).To(Succeed())
+		Expect(os.Unsetenv("COMPOSER_VENDOR_DIR")).To(Succeed())
+	})
+
+	context("when composer.json is present", func() {
+		it.Before(func() {
+			Expect(ioutil.WriteFile(filepath.Join(workingDir, "composer.json"), []byte("{}"), 0644)).NotTo(HaveOccurred())
+		})
+
+		it(`requires "composer" and "php" and provides "composer-packages"`, func() {
+			detectResult, err := detect(packit.DetectContext{WorkingDir: workingDir})
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(detectResult.Plan).To(Equal(packit.BuildPlan{
+				Provides: []packit.BuildPlanProvision{
+					{
+						Name: composer.ComposerPackagesDependency,
+					},
+				},
+				Requires: []packit.BuildPlanRequirement{
+					{
+						Name: "composer",
+						Metadata: composer.BuildPlanMetadata{
+							Build: true,
+						},
+					},
+					{
+						Name: "php",
+						Metadata: composer.BuildPlanMetadata{
+							Build: true,
+						},
+					},
+				},
+			}))
+
+			Expect(phpVersionResolver.ResolveCall.Receives.ComposerJsonPath).To(Equal(filepath.Join(workingDir, "composer.json")))
+			Expect(phpVersionResolver.ResolveCall.Receives.ComposerLockPath).To(Equal(filepath.Join(workingDir, "composer.lock")))
+		})
+
+		context("when PhpVersionResolver returns values", func() {
+			it.Before(func() {
+				phpVersionResolver.ResolveCall.Returns.Version = "php-version-from-resolver"
+				phpVersionResolver.ResolveCall.Returns.VersionSource = "php-version-source-from-resolver"
+			})
+
+			it(`requires "php" with version and version-source metadata`, func() {
+				detectResult, err := detect(packit.DetectContext{WorkingDir: workingDir})
+				Expect(err).NotTo(HaveOccurred())
+
+				Expect(detectResult.Plan).To(Equal(packit.BuildPlan{
+					Provides: []packit.BuildPlanProvision{
+						{
+							Name: composer.ComposerPackagesDependency,
+						},
+					},
+					Requires: []packit.BuildPlanRequirement{
+						{
+							Name: "composer",
+							Metadata: composer.BuildPlanMetadata{
+								Build: true,
+							},
+						},
+						{
+							Name: "php",
+							Metadata: composer.BuildPlanMetadata{
+								Build:         true,
+								Version:       "php-version-from-resolver",
+								VersionSource: "php-version-source-from-resolver",
+							},
+						},
+					},
+				}))
+			})
+		})
+
+		context("when composer.lock is not present", func() {
+			it("will log a warning", func() {
+				_, err := detect(packit.DetectContext{WorkingDir: workingDir})
+				Expect(err).NotTo(HaveOccurred())
+
+				Expect(buffer).To(ContainLines("WARNING: Include a 'composer.lock' file with your application! This will make sure the exact same version of dependencies are used when you build. It will also enable caching of your dependency layer."))
+			})
+		})
+
+		context("failure cases", func() {
+			it("will return an error from PhpVersionResolver", func() {
+				phpVersionResolver.ResolveCall.Returns.Err = errors.New("some error")
```

Comment:
nit: can this go inside an `it.Before` as it's a part of test setup

Reply:
done